source/transactions/transactions.rst

```diff
@@ -1212,9 +1221,72 @@ What happens when a command object passed to RunCommand already contains a trans
 Applications should not run such commands inside a transaction.
 
+Majority write concern is used when retrying commitTransaction
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Drivers should apply a majority write concern when retrying commitTransaction to
+guard against a transaction being applied twice.
+
+Consider the following scenario:
+
+1. The driver is connected to a replica set where node A is primary.
+2. The driver sends commitTransaction to A with ``w:1``. A commits the
+   transaction but dies before it can reply. This constitutes a retryable error,
+   which means the driver can retry the commitTransaction command.
+3. Node B is briefly elected.
+4. The driver retries commitTransaction on B with ``w:1``, and B replies with a
+   NoSuchTransaction error code and TransientTransactionError error label. This
+   implies that the driver may retry the entire transaction.
+5. Node A revives before B has done any ``w:majority`` writes and is reëlected
+   as primary.
+6. The driver then retries the entire transaction on A where it commits
+   successfully.
+
+The outcome of this scenario is that two complete executions of the transaction
+operations are permanently committed on node A.
+
+Drivers can avoid this scenario if they always use a majority write concern when
+retrying commitTransaction. Applying a majority write concern to step four in
+the above scenario would lead to one of the following possible outcomes:
+
+- Node B replies with failed response, which does not include a
+  TransientTransactionError error label. This does not constitute a retryable
+  error. Control is returned to the user.
+- Node B replies with a successful response (e.g. ``ok:1``) indicating that the
+  retried commitTransaction has succeeded durably and the driver can continue.
+  Control is returned to the user.
+- Node B replies with a wtimeout error. This does not constitute a retryable
+  error. Control is returned to the user.
+- Node B replies with a failure response that includes the
+  TransientTransactionError label, which indicates it is safe to retry the
+  entire transaction. Drivers can trust that a server response will not include
+  both a write concern error and TransientTransactionError label (see:
+  `SERVER-37179 <https://jira.mongodb.org/browse/SERVER-37179>`_). 
+
+Adding a majority write concern only when retrying commitTransaction provides a
+good compromise of performance and durability. Applications can use ``w:1`` for
+the initial transaction attempt for a performance advantage in the happy path.
+In the event of retryable error, the driver can upgrade commitTransaction to use
+``w:majority`` and provide additional guarantees to the user and avoid any risk
+of committing the transaction twice. Note that users may still be vulnerable to
+rollbacks by using ``w:1`` (as with any write operation).
+
+While it's possible that the original write concern may provide greater
```

Comment:
@ShaneHarvey: Justification for overriding `w` with "majority" is here.

Reply:
Thanks! Is there still a chance for the txn to be reapplied with `{w: "majority", j: False}`? If so, I think we should also override "j".

Reply:
On second thought, I think we should omit overriding `j`. In the event that the server is running with journaling disabled the write concern might never be satisfiable. If the user has explicitly provided a `j` value in their original write concern, I think we should leave it as-is. Otherwise, we should rely on `writeConcernMajorityJournalDefault`. If `writeConcernMajorityJournalDefault`, we should trust that it was done so for good reason (again, perhaps journaling is disabled) and not interfere with it.